File: mobile/test/tour_line_test.dart
import 'dart:math' as math;

import 'package:aetherride_mobile/domain/tours/tour_line.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('fitTourLine needs two points', () {
    expect(fitTourLine(const []), isNull);
    expect(
      fitTourLine(const [
        [8.6, 49.4],
      ]),
      isNull,
    );
  });

  test('open line is not a loop and stays in pad', () {
    final line = fitTourLine(const [
      [8.6, 49.4],
      [8.7, 49.5],
    ]);
    expect(line, isNotNull);
    expect(line!.loop, isFalse);
    expect(line.points.length, 2);
    expect(line.start.x, inInclusiveRange(8, 56));
    expect(line.end.y, inInclusiveRange(8, 56));
  });

  test('closed ring is a loop', () {
    final loop = fitTourLine(const [
      [8.68, 49.4],
      [8.7, 49.41],
      [8.72, 49.4],
      [8.7, 49.39],
      [8.68, 49.4],
    ]);
    expect(loop, isNotNull);
    expect(loop!.loop, isTrue);
    expect(loop.points.length, 5);
  });

  test('downsample keeps ends', () {
    final many = [
      for (var i = 0; i < 400; i++)
        [8.6 + i * 0.001, 49.4 + math.sin(i / 12) * 0.01],
    ];
    final sampled = downsampleLngLats(many, 64);
    expect(sampled.length, 64);
    expect(sampled.first, many.first);
    expect(sampled.last, many.last);
  });

  test('aspect: north-south vs east-west', () {
    final ns = fitTourLine(const [
      [8.7, 49.3],
      [8.7, 49.5],
    ]);
    final ew = fitTourLine(const [
      [8.6, 49.4],
      [8.9, 49.4],
    ]);
    expect((ns!.start.x - ns.end.x).abs(), lessThan(1));
    expect((ew!.start.y - ew.end.y).abs(), lessThan(1));
    expect((ns.end.y - ns.start.y).abs(), greaterThan(20));
    expect((ew.end.x - ew.start.x).abs(), greaterThan(20));
  });

  test('wide box still padded', () {
    final wide = fitTourLine(
      const [
        [8.6, 49.4],
        [8.8, 49.45],
        [8.7, 49.5],
      ],
      width: 128,
      height: 64,
    );
    expect(wide, isNotNull);
    expect(wide!.start.x, inInclusiveRange(8, 120));
    expect(wide.start.y, inInclusiveRange(7, 57));
  });

  test('trackCoordsOf prefers merged coordinates', () {
    expect(
      trackCoordsOf(
        coordinates: const [
          [1.0, 2.0],
        ],
        tour: const [
          [3.0, 4.0],
          [5.0, 6.0],
        ],
      ),
      [
        [3.0, 4.0],
        [5.0, 6.0],
      ],
    );
    expect(
      trackCoordsOf(
        coordinates: const [
          [1.0, 2.0],
          [3.0, 4.0],
        ],
        tour: const [
          [9.0, 9.0],
          [8.0, 8.0],
        ],
      ),
      [
        [1.0, 2.0],
        [3.0, 4.0],
      ],
    );
  });
}

File: src/lib/tours/tourLine.test.ts
/**
 * npx tsx src/lib/tours/tourLine.test.ts
 */
import assert from "node:assert/strict";
import { downsampleLngLats, fitTourLine } from "./tourLine";

assert.equal(fitTourLine([]), null);
assert.equal(fitTourLine([[8.6, 49.4]]), null);

const line = fitTourLine([
  [8.6, 49.4],
  [8.7, 49.5],
]);
assert.ok(line);
assert.ok(line.d.startsWith("M"));
assert.ok(line.d.includes(" L"));
assert.equal(line.loop, false);
assert.equal(line.pointCount, 2);
assert.ok(line.start.x >= 8 && line.start.x <= 56);
assert.ok(line.end.y >= 8 && line.end.y <= 56);

const loop = fitTourLine([
  [8.68, 49.4],
  [8.7, 49.41],
  [8.72, 49.4],
  [8.7, 49.39],
  [8.68, 49.4],
]);
assert.ok(loop);
assert.equal(loop.loop, true);
assert.equal(loop.pointCount, 5);

const many: [number, number][] = [];
for (let i = 0; i < 400; i++) {
  many.push([8.6 + i * 0.001, 49.4 + Math.sin(i / 12) * 0.01]);
}
const sampled = downsampleLngLats(many, 64);
assert.equal(sampled.length, 64);
assert.deepEqual(sampled[0], many[0]);
assert.deepEqual(sampled[63], many[399]);

const ns = fitTourLine([
  [8.7, 49.3],
  [8.7, 49.5],
]);
const ew = fitTourLine([
  [8.6, 49.4],
  [8.9, 49.4],
]);
assert.ok(ns && ew);
assert.ok(Math.abs(ns.start.x - ns.end.x) < 1, "north-south stays centered in x");
assert.ok(Math.abs(ew.start.y - ew.end.y) < 1, "east-west stays centered in y");
assert.ok(Math.abs(ns.end.y - ns.start.y) > 20);
assert.ok(Math.abs(ew.end.x - ew.start.x) > 20);

const wide = fitTourLine(
  [
    [8.6, 49.4],
    [8.8, 49.45],
    [8.7, 49.5],
  ],
  128,
  64,
);
assert.ok(wide);
assert.ok(wide.start.x >= 8 && wide.start.x <= 120);
assert.ok(wide.start.y >= 7 && wide.start.y <= 57);

console.log("tourLine.test.ts ok");
